biz.aQute.bndlib/src/aQute/bnd/cdi/AnnotationReader.java

```diff
@@ -0,0 +1,340 @@
+					break;
+				}
+				definitions.get(0).serviceOrigin = ElementType.TYPE_USE;
+				int targetIndex = annotation.getTargetIndex();
+				if (targetIndex == Annotation.TARGET_INDEX_EXTENDS) {
+					definitions.get(0).service.add(extendsClass);
+				} else {
+					definitions.get(0).service.add(interfaces[targetIndex]);
+				}
+				break;
+			case TYPE :
+				if (definitions.get(0).serviceOrigin != null
+					&& definitions.get(0).serviceOrigin == ElementType.TYPE_USE) {
+					analyzer.error("@Service cannot be used both on TYPE and TYPE_USE on class %s", clazz);
+					break;
+				}
+				definitions.get(0).serviceOrigin = ElementType.TYPE;
+				Object[] serviceClasses = annotation.get("value");
+
+				if (serviceClasses != null && serviceClasses.length > 0) {
+					for (Object serviceClass : serviceClasses) {
+						definitions.get(0).service.add((TypeRef) serviceClass);
+					}
+				} else if (interfaces != null && interfaces.length > 0) {
+					for (TypeRef inter : interfaces) {
+						definitions.get(0).service.add(inter);
+					}
+				} else {
+					definitions.get(0).service.add(clazz.getClassName());
+				}
+				break;
+			default :
+				// this can't happen, @Service is limited to the above targets
+		}
+	}
+
+	@Override
+	public void classBegin(int access, TypeRef name) {
+		definitions.get(0).implementation = name;
+
+		PackageDef packageDef = packageInfos.computeIfAbsent(name.getPackageRef(), k -> {
+			Clazz packageInfoClazz = analyzer.getPackageInfo(k);
+			if (packageInfoClazz != null) {
+				try {
+					PackageDef pd = new PackageDef();
+					packageInfoClazz.parseClassFileWithCollector(pd);
+					return pd;
+				} catch (Exception e) {
+					analyzer.exception(e, "Error while processing package-info of class %s", clazz);
+				}
+			}
+			return new PackageDef();
+		});
+
+		if (packageDef.marked != null) {
+			definitions.get(0).marked = packageDef.marked.matches(name.getFQN());
+		}
+	}
+
+	@Override
+	public void classEnd() throws Exception {
+		member = null;
+		referenceDef = null;
+	}
+
+	@Override
+	public void extendsClass(TypeRef name) {
+		this.extendsClass = name;
+	}
+
+	@Override
+	public void field(FieldDef field) {
+		this.member = field;
+	}
+
+	@Override
+	public void implementsInterfaces(TypeRef[] interfaces) {
+		this.interfaces = interfaces;
+	}
+
+	@Override
+	public void memberEnd() {
+		member = null;
+		referenceDef = null;
+	}
+
+	@Override
+	public void method(MethodDef method) {
+		if (method.isAbstract() || method.isBridge())
+			return;
+
+		if (!baseclass && method.isPrivate())
+			return;
+
+		this.member = method;
+
+		if (member instanceof MethodDef) {
+			String signature = member.getSignature();
+			if (signature == null) {
+				signature = method.getDescriptor()
+					.toString();
+			}
+			MethodSignature sig = MethodSignature.of(signature);
+			if (sig.parameterTypes.length == 1) {
+				JavaTypeSignature javaTypeSignature = sig.parameterTypes[0];
```

Comment:
Use a MethodResolver to get the parameter. See 
https://github.com/bndtools/bnd/blob/ceba0425550d6bd6bfb572e5419488cc042ddfd9/biz.aQute.bndlib/src/aQute/bnd/component/AnnotationReader.java#L1122
https://github.com/bndtools/bnd/blob/ceba0425550d6bd6bfb572e5419488cc042ddfd9/biz.aQute.bndlib/src/aQute/bnd/component/AnnotationReader.java#L1133

This will nicely resolve any indirection through a type parameter. 

To create the ClassSignature for the type see:
https://github.com/bndtools/bnd/blob/ceba0425550d6bd6bfb572e5419488cc042ddfd9/biz.aQute.bndlib/src/aQute/bnd/component/AnnotationReader.java#L163-L164